### PR TITLE
RM-140, RM-138 and RM-284

### PR DESCRIPTION
**Pull request by nallana**
> [!NOTE]
> **Medium Risk**
> Touches GPU fusion execution, WGSL generation, and VM variable-store behavior; regressions could manifest as wrong results or residency leaks/clears across accelerated paths. Includes fallbacks and extensive new tests, but changes are broad across runtime/accel layers.
> 
> **Overview**
> Adds **multi-output fused elementwise execution**: fusion planning now tracks assignable intermediate results, generates a single WGSL kernel that writes multiple outputs, and executes it via a new `AccelProvider::fused_elementwise_multi` WGPU implementation (with per-output fallback to `fused_elementwise`).
> 
> Introduces a **residency mark hook** (`register_residency_mark`/`mark_residency`) and wires it through `runmat-accelerate`, runtime GPU helpers, and builtins so GPU results are consistently marked resident; also avoids clearing residency when overwriting a variable with the same GPU handle.
> 
> Extends fused WGSL builtin support/semantics (`hypot` implementation; `asinh`/`acosh`/`atanh`, `sign`, `fix`, `pow2`, `mod`, `rem` with Inf/NaN edge cases) and adjusts omitnan `sum` to fall back when device cleanup isn’t supported. Updates telemetry assertions and adds new end-to-end regression tests covering fusion barriers, inline casts, and the new builtin semantics.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ad8e380053c93f6d08913de4a29d96137d8e54f9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: crates/runmat-accelerate/src/fusion.rs [hunk 1]
```diff
@@ -2628,13 +2657,36 @@ fn builtin_expr(
...
+            ));
+        }
         "sin" => "sin",
         "cos" => "cos",
         "tan" => "tan",
         "asin" => "asin",
         "acos" => "acos",
         "atan" => "atan",
         "atan2" => return builtin_binary("atan2", inputs, exprs),
+        "hypot" => return builtin_binary("hypot", inputs, exprs),
```
[the remark this answers is not in the document]
Comment (gneeri) [reply crates/runmat-accelerate/src/fusion.rs:2853]: @nallana, for sure this needs to be dealt with. Let me know if you don't have time and I'll work through it.

File: crates/runmat-accelerate/src/fusion.rs [hunk 2]
```diff
@@ -2652,6 +2704,24 @@ fn builtin_expr(
...
+                    Some(format!(
+                        "log({arg} + sqrt(({arg} - {one}) * ({arg} + {one})))"
+                    ))
+                }
+                "atanh" => {
+                    let arg = exprs.get(inputs.first()?).cloned()?;
+                    let one = cast_literal(scalar_ty, "1.0");
+                    let half = cast_literal(scalar_ty, "0.5");
+                    Some(format!("({half} * log(({one} + {arg}) / ({one} - {arg})))"))
+                }
```
[the remark this answers is not in the document]
Comment (gneeri) [reply crates/runmat-accelerate/src/fusion.rs]: good to have

File: crates/runmat-accelerate/src/fusion.rs [hunk 3]
```diff
@@ -2628,13 +2657,36 @@ fn builtin_expr(
         "isinf" => return builtin_unary_call("isInf", inputs, exprs),
         "isnan" => return builtin_unary_call("isNan", inputs, exprs),
         "single" | "double" | "gpuarray" => return builtin_identity(inputs, exprs),
+        "fix" => return builtin_unary_call("trunc", inputs, exprs),
+        "sign" => return builtin_unary_call("sign", inputs, exprs),
+        "mod" => {
+            let lhs = exprs.get(inputs.first()?).cloned()?;
```
Comment (gneeri) [inline crates/runmat-accelerate/src/fusion.rs:2830]: When rhs = -Inf and lhs is finite positive, isInf(rhs) && isFinite(lhs) is true, so WGSL returns lhs (e.g., 5). But MATLAB returns -Inf for mod(5, -Inf) because floor(5 / -Inf) = floor(-0) = 0 and then the sign-correction step kicks in. The mod_real_parity_matrix_matches_expected_values test doesn't include the (positive, -Inf) or (negative, -Inf) cases. This edge case may not matter for typical workloads, but it's a spec divergence worth tracking.

The rem_real_expected test helper at line ~1762 does handle b.is_infinite() && a.is_finite() correctly by returning a, which matches rem's MATLAB semantics. mod diverges from MATLAB here.

**Review by gneeri — CHANGES_REQUESTED** [review]
Cursorbot fixes should be made, as well as the other noted changes.

File: crates/runmat-accelerate/src/fusion.rs (hunk 3 above)
Comment (gneeri) [reply crates/runmat-accelerate/src/fusion.rs:2830]: done

**Review by gneeri — APPROVED** [review]
